examples/quickrefl/core/editors/importdatawidgets/famousloader/importtextview.cpp

```diff
@@ -105,25 +84,52 @@ void ImportTextView::highlightCurrentLine()
     setExtraSelections(extraSelections);
 }
 
-//![cursorPositionChanged]
+//! Get the first visible block id
+int ImportTextView::getFirstVisibleBlockId()
+{
+    QTextCursor curs = QTextCursor(this->document());
+    curs.movePosition(QTextCursor::Start);
+    for(int i=0; i < this->document()->blockCount(); ++i)
+    {
+        QTextBlock block = curs.block();
 
-//![extraAreaPaintEvent_0]
+        QRect r1 = this->viewport()->geometry();
+        QRect r2 = this->document()->documentLayout()->blockBoundingRect(block).translated(
+                    this->viewport()->geometry().x(), this->viewport()->geometry().y() - (
+                        this->verticalScrollBar()->sliderPosition()
+                        ) ).toRect();
+
+        if (r1.contains(r2, true)) { return i; }
+
+        curs.movePosition(QTextCursor::NextBlock);
+    }
+
+    return 0;
+}
 
 void ImportTextView::lineNumberAreaPaintEvent(QPaintEvent *event)
 {
     QPainter painter(lineNumberArea);
     painter.fillRect(event->rect(), Qt::lightGray);
 
-//![extraAreaPaintEvent_0]
+    int blockNumber = this->getFirstVisibleBlockId();
+    QTextBlock block = this->document()->findBlockByNumber(blockNumber);
+    QTextBlock prev_block = (blockNumber > 0) ? this->document()->findBlockByNumber(blockNumber-1) : block;
+    int translate_y = (blockNumber > 0) ? -this->verticalScrollBar()->sliderPosition() : 0;
+    int top = this->viewport()->geometry().top();
+    int additional_margin;
+    if (blockNumber == 0)
+        // Simply adjust to document's margin
+        additional_margin = (int) this->document()->documentMargin() -1 - this->verticalScrollBar()->sliderPosition();
+    else
+        // Getting the height of the visible part of the previous "non entirely visible" block
+        additional_margin = (int) this->document()->documentLayout()->blockBoundingRect(prev_block)
+                .translated(0, translate_y).intersected(this->viewport()->geometry()).height();
 
-//![extraAreaPaintEvent_1]
-    QTextBlock block = firstVisibleBlock();
-    int blockNumber = block.blockNumber();
-    int top = qRound(blockBoundingGeometry(block).translated(contentOffset()).top());
-    int bottom = top + qRound(blockBoundingRect(block).height());
-//![extraAreaPaintEvent_1]
+    // Shift the starting point
+    top += additional_margin;
+    int bottom = top + (int) this->document()->documentLayout()->blockBoundingRect(block).height();
```

Comment:
Since you are using such kind of access everywhere, consider introducing a method
```
QLayout* documentLayout() { return document()->documentLayout());
```